packages/api/endpoints/executions.js

```diff
@@ -45,7 +46,50 @@ async function get(req, res) {
   }
 }
 
+/**
+ * Delete an execution
+ *
+ * @param {Object} req - express request object
+ * @param {Object} res - express response object
+ * @returns {Promise<Object>} the promise of express response object
+ */
+async function deleteExecution(req, res) {
+  const {
+    executionModel = new models.Execution(),
+    executionPgModel = new ExecutionPgModel(),
+    knex = await getKnexClient(),
+  } = req.testContext || {};
+
+  const { arn } = req.params;
+
+  let existingExecution;
+  try {
+    existingExecution = await executionModel.get({ arn });
+  } catch (error) {
+    if (error instanceof RecordDoesNotExist) {
+      return res.boom.notFound('No record found');
+    }
+    throw error;
+  }
+
+  try {
+    await knex.transaction(async (trx) => {
+      await executionPgModel.delete(trx, { arn });
+      await executionModel.delete({ arn });
+    });
+    return res.send({ message: 'Record deleted' });
+  } catch (error) {
+    // Delete is idempotent, so there may not be a DynamoDB
+    // record to recreate
+    if (existingExecution) {
+      await executionModel.create(existingExecution);
+    }
+    throw error;
+  }
+}
```

Comment:
@botanical had a good question here along the lines of "Why?" 😄 

I took this logic from https://github.com/nasa/cumulus/pull/2245/files#diff-2e28b62d59bb9e81dcbe3b0f48929ad59c25e359b98cc9d67571d96c3e3123f5 because I think the idea is that we could have deleted the Dynamo record in the transaction and have the transaction still fail for _reasons_ after. That, or I'm misunderstanding how transactions fundamentally work and _both_ deletes will execute even if the first one fails. I don't think that's right but need to verify why we're re-creating the Dynamo record like this.

Reply:
Wrote a bunch of commentary before seeing this comment ^      See above.

Reply:
The pattern of "recreate the record" if the transaction fails is not necessary for the standard PG/dynamo delete case. In that case, if either delete fails, the other doesn't happen.

The pattern **is necessary** once you add a third operation to the transaction, such as delete from Elasticsearch (which I'm doing as part of 2208). Because in that case you could have:

- delete from PG succeeds
- delete from Dynamo succeeds
- delete from Elasticsearch errors/throws

However, the Dynamo delete wouldn't "roll back" when the Elasticsearch delete errors/throws, so I added logic to manually recreate the Dynamo record in those cases. Otherwise the three systems would be out of sync. It's a bit unfortunate but a necessary evil while we are temporarily maintaining data across three systems and ultimately gives us a transactional guarantee about the state of data between them. 

So this pattern isn't necessary until we add writing to/deleting from Elasticsearch to the endpoint logic, which doesn't exist in this PR.

Reply:
AH! That makes sense. We were thinking there was a good reason but couldn't find it 😄.

It would make sense to me to leave this as-is on both the async operations PR and this one for consistency and to support ES down the line. Unless it creates duplicate Dynamo records 🤔. Is there a chance we could fail the PG delete, not attempt the Dynamo delete, and end up creating a duplicate Dynamo record in the `catch`? I could write a test to try it.

Reply:
I would suggest going the other way: remove the extra try/catch and recreation of the dynamo record. At best, it's confusing to leave it in when we don't need it yet. And at worst, it may have some unintended side effects if it's left in, because the Dynamo recreate will **overwrite** the non-deleted Dynamo record.

Reply:
Ah! Nope, I missed this. I think I saw the 👍 on Mark's response and thought I already hit it.

Here's the change to remove the try/catch and an added test for trying to delete an execution that doesn't exist. I think it makes sense to leave the Dynamo lookup before trying to delete so we can return that 404. Alternatively we might be able to look at the number of rows affected by the delete and if it's 0 say the record couldn't be found. I'm not totally sure that's a good pattern though. It might be possible that a record fails to delete due to a fk constraint and, in that case, we wouldn't want to return a 404.

https://github.com/nasa/cumulus/pull/2252/commits/91eebde2c710bb63f3d28d0829db592f2a3f0018

Reply:
👍🏻  looks good.